Limit request size

danielmitterdorfer opened the pull request:
With this PR we add the possibility to limit the size of requests on multiple levels:

* All in-flight requests on HTTP level, configured by `http.breaker.request.limit` (default 128 MB)
* All in-flight requests on transport level, configured by `transport.breaker.request.limit` (default 128 MB)
* Individual outgoing requests on transport level, configured by `transport.outgoing.request.limit`

If a limit is breached on transport level, we throw `CircuitBreakingException`, if a limit is breached on HTTP level we return HTTP 413 (request entity too large).

Closes #16011

danielmitterdorfer commented:
For the review I am specifically interested in three things:

1. Should configuration properties a standard naming scheme? Currently I have adhered to the naming scheme in the respective implementation classes where the properties are defined. Note that for the circuit breaker based properties we also have the properties `transport.breaker.request.overhead` and `http.breaker.request.overhead`.
2. Are the default values ok? I have basically derived the values based on the fact that `http.max_content_length` is already 100 MB by default and it would not make sense to put a lower default limit on all in-flight requests.
3. Should we adjust the default value `indices.breaker.total.limit` for the parent circuit breaker given that we have now introduced two new child breakers?

danielmitterdorfer commented:
@dakrone: Do you have time to review this PR?

Dieken commented on core/src/main/java/org/elasticsearch/http/netty/RequestSizeLimitHandler.java:
```diff
@@ -0,0 +1,86 @@
+                // on some occasions, we will get 0 (e.g. when the message is chunked and we get the first chunk)
+                if (reservation > 0) {
+                    logger.debug("Increase amount of in-flight requests by [{}] bytes on channel id [{}]", reservation, channelId);
+                    assert reservationsPerChannel.containsKey(channelId) == false : "Should not contain multiple reservations per channel";
```
I feel this is problematic,  if HTTP pipe-lining or chunked transferring is enabled,  handleUpstream() will be called multiple times but handleDownstream() may not get called because the response isn't ready yet,  then the assertion will fail or the entry in reservationsPerChannel is overwritten if JVM assertion is disabled.

danielmitterdorfer replied:
That's a good catch. I have changed the implementation now to handle also pipelined requests.

Dieken commented on core/src/main/java/org/elasticsearch/http/netty/RequestSizeLimitHandler.java:
```diff
@@ -0,0 +1,86 @@
+                ChannelBuffer buf = ((HttpMessage) message).getContent();
+                int channelId = ctx.getChannel().getId();
+                int reservation = buf.readableBytes();
+                // on some occasions, we will get 0 (e.g. when the message is chunked and we get the first chunk)
```
The comment here seems wrong,  this handler is after the HTTPChunkHandler instance, so it never gets chunked message as "HttpMessage" here.  But indeed the "HttpMessage" may have empty body,  for example,  HTTP GET requests.

danielmitterdorfer replied:
Yes, this was an old one. I've removed it as I think it isn't too helpful anyway.

Dieken commented:
Maybe name the in-flight requests related settings to ".....requests....."  or even "......inflight_requests....".

Dieken commented on core/src/main/java/org/elasticsearch/http/netty/RequestSizeLimitHandler.java:
```diff
@@ -0,0 +1,86 @@
+
+    private final Limiter limiter;
+
+    private final Map<Integer, Reservation> reservationsPerChannel = new ConcurrentHashMap<>();
```
Is AttributeKey (http://netty.io/4.0/api/io/netty/channel/ChannelHandler.html) cheaper than ConcurrentHashMap?

The booking in my patch https://github.com/Dieken/elasticsearch/commit/f2d487eaf1213daa6ad25f13fbccdd1d6b5930f4 is cheaper but it sits in higher level code.

Maybe introduce similar trick like OrderedUpstreamMessageEvent and OrderedDownstreamMessageEvent,  let handleDownstream() know the reservation from its argument "e".

danielmitterdorfer replied:
Regarding `AttributeKey`: I would rather rely on JDK classes as long as others don't provide a significant improvement. You're right that using `LongAdder` is cheaper but on a higher level you already get request / response correlation basically for free.

The idea to piggyback on the message event is nice but I'd also avoid tying reservations together with the events.

Dieken commented on core/src/main/java/org/elasticsearch/common/limit/CountingLimiter.java:
```diff
@@ -0,0 +1,59 @@
+    @Override
+    public Reservation addBytesChecked(long numBytes) {
+        // check for overflows
+        long newBytesUsed = Math.addExact(bytesUsed, numBytes);
```
Math.addExact() is introduced in JDK 8.

danielmitterdorfer replied:
Nice catch. I'll watch out in the backport but I'd keep it on master.

Dieken commented on core/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java:
```diff
@@ -64,11 +65,27 @@
-
+    public static final Setting<ByteSizeValue> TRANSPORT_CIRCUIT_BREAKER_LIMIT_SETTING =
+        Setting.byteSizeSetting("transport.breaker.request.limit",
+            new ByteSizeValue(128, ByteSizeUnit.MB), Property.Dynamic, Property.NodeScope);
```
http.max_content_length is 100MB by default,  maybe we should allow at least two concurrent requests :-)

danielmitterdorfer replied:
Agreed. I've adjusted the default to 200MB now (also for HTTP) and renamed the breakers to ...requests_inflight...

Dieken commented on core/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java:
```diff
@@ -64,11 +65,27 @@
+
+    public static final Setting<ByteSizeValue> HTTP_CIRCUIT_BREAKER_LIMIT_SETTING =
+        Setting.byteSizeSetting("http.breaker.request.limit",
+            new ByteSizeValue(128, ByteSizeUnit.MB), Property.Dynamic, Property.NodeScope);
```
http.max_content_length * 2  ??

danielmitterdorfer replied:
see above.

danielmitterdorfer commented:
@Dieken: Thanks for your comments. I am currently working on another issue but I am having a closer look ASAP.

danielmitterdorfer commented:
Regarding the HTTP status code 413: Here is the relevant part from [RFC 2616, section 10](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html)

> 10.4.14 413 Request Entity Too Large
> 
> The server is refusing to process a request because the request entity is larger than  the server is willing or able to process. The server MAY close the connection to prevent the client from continuing the request.

> If the condition is temporary, the server SHOULD include a Retry- After header field to indicate that it is temporary and after what time the client MAY try again.

So we can add an `Retry-After` header (see also [RFC 2616, section 14](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html)) with a relative delay of - say - 10 seconds to indicate that this condition may be temporary. Note that the condition is only temporary in case of multiple in-flight requests. If a single request hits the limit, it is still permanent (until somebody changes the setting of course). It may be even possible to detect these two cases but I am not sure it is worth the effort and additional code complexity.

dakrone commented on core/src/main/java/org/elasticsearch/common/limit/LimitingStreamOutput.java:
```diff
@@ -0,0 +1,87 @@
+
+    @Override
+    public void reset() throws IOException {
+        addBytesChecked(-bytesWritten);
```
Does `reset()` actually "free" the bytes that have been written so far?

danielmitterdorfer replied:
No, this does not free the bytes but I still think it is necessary to reduce the number of reserved bytes. Consider this scenario:

```java
LimitingStreamOutput stream = new LimitingStreamOutput(new BytesStreamOutput());
stream.writeByte((byte) 0);
stream.writeByte((byte) 0);
stream.writeByte((byte) 0); // at this point we've written (and reserved) 3 bytes
stream.reset();  // to my understanding we should now reset the number of reserved bytes to 0
stream.writeByte((byte) 0); // reserve 1 byte
stream.close();  // return all reserved bytes
```

Wdyt?

dakrone replied:
Looking at the implementation:

```java
    @Override
    public void reset() {
        // shrink list of pages
        if (bytes.size() > BigArrays.PAGE_SIZE_IN_BYTES) {
            bytes = bigarrays.resize(bytes, BigArrays.PAGE_SIZE_IN_BYTES);
        }

        // go back to start
        count = 0;
    }
```

It doesn't necessarily "free" bytes, just resizes them back to a 16384 byte array if more than 16384 bytes have been written already, so I'm not sure what to do here

danielmitterdorfer replied:
As discussed I think both approaches are ok. If we're not reducing the amount in `#reset()` we're a bit more conservative (as we keep the count although there may not be more memory allocated) but when `#close()` is called we'll always free all written bytes, even if `#reset()` has been called in between. So I'll change the implementation and don't reduce the count in `#reset()` anymore.

dakrone commented on core/src/main/java/org/elasticsearch/common/limit/Reservation.java:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * Returns the reservation, i.e. it reduces the number of reserved bytes on the correct limiter.
+     */
+    public void doReturn() {
```
Hmm.. what about making `Reservation` implement `Closeable` and then releasing them in the `close()` method? This would allow them to be used in try-with-resources

danielmitterdorfer replied:
I like that idea. I'll do that. Even if it is not used in a try-with-resources block it makes the intention clearer.

dakrone commented on core/src/main/java/org/elasticsearch/http/netty/RequestSizeLimitHandler.java:
```diff
@@ -0,0 +1,112 @@
+            if (message instanceof HttpMessage) {
+                ChannelBuffer buf = ((HttpMessage) message).getContent();
+                int channelId = ctx.getChannel().getId();
+                Tuple<Integer, Integer> key = key(ctx, e);
```
I believe since this is high-traffic code, you could put the `key` creating inside of the `if (reservation > 0)` statement so it isn't created unless we're actually adding bytes to help with GC

danielmitterdorfer replied:
Sure, you're totally right.

dakrone commented on core/src/main/java/org/elasticsearch/http/netty/RequestSizeLimitHandler.java:
```diff
@@ -0,0 +1,112 @@
+            if (reservationsPerChannel.containsKey(key)) {
+                Reservation reservation = reservationsPerChannel.remove(key);
+                int channelId = ctx.getChannel().getId();
+                logger.debug("Decrease amount of in-flight requests by [{}] bytes on channel id [{}]",
```
this should probably be `trace` level right? this could be a lot of logging

danielmitterdorfer replied:
I thought `debug` would be low-level enough. I'll change it to `trace`.

dakrone commented on core/src/main/java/org/elasticsearch/transport/local/LocalTransport.java:
```diff
@@ -223,7 +238,7 @@ public void sendRequest(final DiscoveryNode node, final long requestId, final St
             }
 
-            final byte[] data = stream.bytes().toBytes();
+            final byte[] data = bytesStream.bytes().toBytes();
```
why use `byteStream` here instead of `stream` since `stream` should delegate it correctly anyway?

danielmitterdorfer replied:
`#bytes()` is defined on `BytesStream` but not on `StreamOutput` so `LimitingStreamOutput` does not have this method.

dakrone commented on core/src/main/java/org/elasticsearch/transport/netty/MessageChannelHandler.java:
```diff
@@ -342,4 +313,66 @@ public void onFailure(Throwable e) {
+
+        public Marker(ChannelBuffer buffer) {
+            this.buffer = buffer;
+            this.size = buffer.getInt(buffer.readerIndex() - 4);
```
what's the magical 4 from?

danielmitterdorfer replied:
tbh, I have just moved around the code. There is also another magic number (6) in the enclosing class `MessageChannelHandler` in `messageReceived` which is quite likely related.

bleskes replied:
this 4 is because we just read an int of the buffer in SizeHeaderFrameDecoder  (we should add a comment). I also prefer to have the logic in this class as two utility methods validateRequestIsRead / validateResponseIsRead which get all the info they need. To understand you need to see when the constructor is called anyway. If you feel strongly it should be class, then I'm fine but please doc it.

bleskes replied:
It seems we're keeping this class. Can we add docs to the different fields and what they mean? what is "markedReaderIndex" ? what is "expectedReaderIndex"? also I think the java docs of the class can be improved (i.e., capture the request start position and expected length in order to do validation

danielmitterdorfer replied:
Oh, sorry I totally missed your previous comment... . I look at it, add more docs and push another commit.

danielmitterdorfer replied:
I looked whether it makes sense to remove the class and replace it with validation methods. As this class is stateful, we'd need to pass additional parameters to the validation methods which would lead to method signatures with 6 (for requests) and 7 parameters (for responses). I have added constants for all parts of the header and eliminated all magic numbers related to these headers.

dakrone commented on core/src/main/java/org/elasticsearch/transport/netty/NettyTransportChannel.java:
```diff
@@ -81,6 +87,9 @@ public void sendResponse(TransportResponse response) throws IOException {
     @Override
     public void sendResponse(TransportResponse response, TransportResponseOptions options) throws IOException {
+        if (reservation != null) {
+            reservation.doReturn();
```
shouldn't this only be released in the `finally` clause of the following `try`? it seems like it'd be better only to release it after the response was sent

danielmitterdorfer replied:
My rationale was to ensure that we return the reservation at this point no matter what. For example, a few lines below there is:

```java
ReleasableBytesStreamOutput bStream = new ReleasableBytesStreamOutput(transport.bigArrays);
```
which can throw a `CircuitBreakingException` (outside the `try`). I have restructured the code so the stream is created in the `try`.

dakrone commented on core/src/main/java/org/elasticsearch/transport/netty/NettyTransportChannel.java:
```diff
@@ -118,6 +127,9 @@ public void sendResponse(TransportResponse response, TransportResponseOptions op
     @Override
     public void sendResponse(Throwable error) throws IOException {
+        if (reservation != null) {
+            reservation.doReturn();
```
same here, releasing after writing to the channel

danielmitterdorfer replied:
Done.

dakrone commented on core/src/test/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateActionTests.java:
```diff
@@ -42,6 +42,7 @@
 import org.elasticsearch.discovery.Discovery;
 import org.elasticsearch.discovery.DiscoverySettings;
 import org.elasticsearch.discovery.zen.DiscoveryNodesProvider;
+import org.elasticsearch.indices.breaker.NoneCircuitBreakerService;
```
unused import (super minor)

danielmitterdorfer replied:
removed.

dakrone commented on core/src/test/java/org/elasticsearch/http/netty/RequestSizeLimitHandlerTests.java:
```diff
@@ -0,0 +1,342 @@
+import java.nio.charset.StandardCharsets;
+
+public class RequestSizeLimitHandlerTests extends ESTestCase {
+    private static final int REQUEST_SIZE_IN_BYTES = 19;
```
Where does this magical 19 come from? It's hard to tell from the tests that use it

danielmitterdorfer replied:
It's the number of bytes in the `match_all` query. I have made it now more explicit:

```java
    private static final String MATCH_ALL_QUERY = "{ \"match_all\": {} }";
    private static final int REQUEST_SIZE_IN_BYTES = MATCH_ALL_QUERY.getBytes().length;
```

dakrone commented:
@danielmitterdorfer left some comments on this! one question I do have is how much impact this will have on performance of HTTP and transport requests, do you have any idea how large the impact will be?

danielmitterdorfer commented:
Regarding your question on the performance impact: I'll run some benchmarks and post the results.

danielmitterdorfer commented:
I've microbenchmarked 3 scenarios with `RequestSizeLimitHandler`:

1. Handle two upstream and downstream events
2. Handle a upstream event, a pipelined upstream event, a pipelined downstream event, a downstream event
3. Handle a pipeline upstream event, a pipelined downstream event, a pipelined upstream event, a pipelined downstream event

(I did this so we have the same number of operations in all cases)

I benchmarked also the current version of `RequestSizeLimitHandler` against the original one (without pipeline support).

A rough summary (exact number depends on the scenario):

Metric | `RequestSizeLimitHandler` without pipeline support | `RequestSizeLimitHandler` (with pipeline support)
--- | --- | ---
Allocation rate | 56 bytes / op | 79 - 104 bytes / op (scenario dependent)
Average Time | 82 - 113 ns / op | 94 - 119 ns / op

Details can be found in the gist: https://gist.github.com/danielmitterdorfer/871cd62d7e16e45ae79d

It is way harder to microbenchmark the overhead of request size limiting on transport level as the new code is not that isolated as on HTTP level. However, I've run macrobenchmarks before and after the change to check for performance-related changes.

The baseline for the test was the last commit on master that I've merged (d17fd33). I've ran two trials with geonames data against a single node Elasticsearch cluster with default settings. The benchmark first bulk-indexes data via HTTP, does a force-merge, runs 20 warmup iterations of all queries and then 10 timed iterations of these queries.

Metric | baseline (without size limiting) | size limiting enabled
--- | --- | ---
Mean Indexing Throughput [docs/s] | 13930, 12271 | 12775, 13331
99% latency match_all query [ms] | 9.79, 9.93 | 10.00, 6.29
99% latency uncached aggregation query [ms] | 115.75, 134.14 | 139.19, 119.25

I also ran the same benchmark against a two node cluster with default settings to also exercise the transport code:

Metric | baseline (without size limiting) | size limiting enabled
--- | --- | ---
Mean Indexing Throughput [docs/s] | 11119, 11711 | 11725, 11715
99% latency match_all query [ms] | 9.93, 6.29 | 8.31, 8.21
99% latency uncached aggregation query [ms] | 134.14, 116.26 | 118.19, 115.67

So in both cases the changes hide in the run-to-run variance of the benchmarks.

danielmitterdorfer commented:
I had a chat with @bleskes about `RequestSizeLimitHandler` and I am currently looking whether we can implement this on a bit higher level (REST layer).

danielmitterdorfer commented:
I've now implemented in-flight request limiting in `RestController#dispatchRequest()` by just using a try-with-resources block. Request-response correlation is not needed anymore with this approach. Hence we only need to hold state for a single reservation (in the try-with-resources block). I've also documented the need for the HTTP in-flight request limit.

bleskes commented on core/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -159,29 +168,32 @@ public RestFilterChain filterChain(RestFilter executionFilter) {
-                    threadContext.putHeader(key, httpHeader);
-                }
+        int contentLength = request.hasContent() ? request.content().length() : 0;
+        try (Reservation reservation = inflightRequestLimiter.addBytesChecked(contentLength)) {
```
I'm afraid this won't work - you only want to decrement this when someone responds to the channel...  it needs to be built into the RestChannel..

danielmitterdorfer replied:
You're right. Upon inspection of the code it looked like all code paths would produce a synchronous response. I'll push a fix soon that returns the reservation in `NettyHttpChannel` where the actual response is sent.

danielmitterdorfer commented:
@bleskes Your comment should now be addressed. I decided to keep `NettyHttpChannel` free of any limiter / circuit breaker logic and just use it to free the reservation when the response has been sent which makes the class simpler.

bleskes commented:
Thank you @danielmitterdorfer very much for iterating on this. I went through this and I'm concerned about the added abstractions and the amount of moving parts. Here are a couple of suggestions to simplify:
1) Remove the Limiter & Releasable notions. Instead interact directly with the circuit breaker and make the channel release the bytes they need to when people respond through them. We can also make the channel be closable and only allow them to be "closed" once in order to protect against double freeing.
2) We have two in-flight requests circuit breakers, one for HTTP and once for transports. I think we can have one that holds for all requests, regardless of where they came from. (we should have a separate discussion about what the default size should be)
3) Given that fact the transport client is being replaced with a java only client, we can remove the max outgoing request size limit and simplify the code a bit more.
4) The transport service interface now has an extra limitStats method that is only used in testing. I think we can remove it and access the circuitbreaker service directly to check all was freed.

Does this all make sense?

danielmitterdorfer commented:
Thanks for your feedback. I'll incorporate the changes soon.

danielmitterdorfer commented:
@bleskes I've addressed your comments now. Can you have another look please?

bleskes commented on core/src/main/java/org/elasticsearch/http/netty/NettyHttpChannel.java:
```diff
@@ -58,19 +58,22 @@
+    private final OrderedUpstreamMessageEvent orderedUpstreamMessageEvent;
+    private final long reservedBytes;
 
     public NettyHttpChannel(NettyHttpServerTransport transport, NettyHttpRequest request,
```
can we remove this constructor variant and document that you can pass null for orderedUpstreamMessageEvent ?

danielmitterdorfer replied:
Sure. I'll also add Javadocs for the constructor.

bleskes commented on core/src/main/java/org/elasticsearch/http/netty/NettyHttpChannel.java:
```diff
@@ -58,19 +58,22 @@
+        this(transport, request, null, reservedBytes, detailedErrorsEnabled);
+    }
+
+    public NettyHttpChannel(NettyHttpServerTransport transport, NettyHttpRequest request,
```
let's add java docs here, it becomes non trivial :)

bleskes commented on core/src/main/java/org/elasticsearch/http/netty/NettyHttpChannel.java:
```diff
@@ -134,6 +137,7 @@ public void sendResponse(RestResponse response) {
             if (!addedReleaseListener && content instanceof Releasable) {
                 ((Releasable) content).close();
             }
+            transport.inFlightRequestsBreaker.addWithoutBreaking(-reservedBytes);
```
can we make this implement AutoClosable, call the close method from here to do all the clean up and make sure it's only closed once?

danielmitterdorfer replied:
I assume that you want `NettyHttpChannel` to implement `AutoCloseable`. I think that would just make sense if we would use `NettyHttpChannel` in a try-with-resources block. I can however implement `Closeable` and add a precondition check there that we do not close it twice.

bleskes replied:
The reason why I suggested AutoClosable is that I don't want close to be idempotent, but rather throw an exception, see Closable docs:

```
  If the stream is already closed then invoking this method has no effect.
```

I'm fine with not implementing any interface at all and just implement the close() method and call it on sendResponse

danielmitterdorfer replied:
I now throw an `IllegalStateException` in case we attempt to send a response on a closed channel or if we try to close a channel twice.

bleskes commented on core/src/main/java/org/elasticsearch/http/netty/NettyHttpServerTransport.java:
```diff
@@ -498,7 +509,10 @@ protected void exceptionCaught(ChannelHandlerContext ctx, ExceptionEvent e) thro
                 return;
             }
-            if (!NetworkExceptionHelper.isCloseConnectionException(e.getCause())) {
+            if (e.getCause() instanceof CircuitBreakingException) {
```
why do we need special handling for CircuitBreakingException ? REQUEST_ENTITY_TOO_LARGE doesn't feel right because it's not about this single request but rather about the accumulation  of all requests?

danielmitterdorfer replied:
We need the special handling because `CircuitBreakingException` is thrown when the amount in-flight bytes exceeds a limit. [RFC 2616, section 10](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html) specifically states that this can be also used to signal temporary condition and that the server should provide a hint in that case (which we could do). While I agree with you that this may not be 100% correct I think this is the best we can do and clients would then be even able to base their retry logic on this status code which in line with the semantics of RFC 2616. What status code would you suggest in that case?

bleskes replied:
If we think that circuit breaking should have a specific rest status, then we should do it in the CircuitBreakerException not here. To me saying the request too big (for any circuitbreaking, but that can be solved) is misleading. I'd prefer server unavailable.

danielmitterdorfer replied:
I think (in general) that defining protocol level information in an exception mixes concepts and makes the exception classes more complex. I was not aware that we already do this so I agree with you that in the context of our code base we should define the status in the exception. Consequently, I have removed the special case handling for `CircuitBreakingException`.

bleskes commented on core/src/main/java/org/elasticsearch/indices/breaker/CircuitBreakerModule.java:
```diff
@@ -39,7 +39,7 @@ protected void configure() {
             impl = HierarchyCircuitBreakerService.class;
         } else if (type.equals("none")) {
-            impl = NoneCircuitBreakerService.class;
+            impl = StaticCircuitBreakerService.class;
```
why was this needed?

danielmitterdorfer replied:
see below.

bleskes commented on core/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java:
```diff
@@ -64,10 +65,17 @@
 
-
+    public static final Setting<ByteSizeValue> IN_FLIGHT_CIRCUIT_BREAKER_LIMIT_SETTING =
+        Setting.byteSizeSetting("net.breaker.inflight_requests.limit",
```
can we use `network.` as a prefix? also please name the setting IN_FLIGHT_REQUESTS?

@dakrone what do you feel should be the default? Do we want to leave it unbound and let a parent circuit breaker kick in? Also since this is the first non-indices related breaker, do you think we should rename all other settings to have INDICES_ prefix in their name?

danielmitterdorfer replied:
Sure, I'll rename it.

bleskes commented on core/src/main/java/org/elasticsearch/indices/breaker/StaticCircuitBreakerService.java:
```diff
@@ -26,14 +26,30 @@
  * Class that returns a breaker that never breaks
  */
-public class NoneCircuitBreakerService extends CircuitBreakerService {
+public class StaticCircuitBreakerService extends CircuitBreakerService {
```
I *think* I understand why this is needed - in a test you needed to have this based on a custom breaker for the new inflight breaker type. I think this signals a bigger problem. This class currently returns the same breaker for all types. Which used to be fine (for all but stats reporting, which was hard coded to field data), because the logic was "return me something that does nothing anyway". Now it becomes "return me the same breaker regardless of what I asked for". 

I think the right solution is to have the set of breakers predefine and registered on the circuit breaker module and have all the breakers configured to NoopCircuitBreaker . That is a big change.  

Alternatively we can make this class be `SharedCircuitBreakerService` where always pass a single circuit breaker in it's constructor and it's document that all requests (including stats) goes to that one.  We can then instantiate it with the NoopCircuitBreaker in CircuitBreakerModule based on config.

danielmitterdorfer replied:
You're right. When unit-testing the `*ServerTransport` classes I needed to provide a `CircuitBreakerService` but I need more fine-grained control over the behaviour of the `CircuitBreaker` that is returned (note that in these tests I just need one breaker). In all integration tests (which use Guice) we still have the old logic of returning always a `NoopCircuitBreaker`.

I don't think that we need all the registration stuff in the circuit breaker module but we should instead define more specifically which breaker implementation we want to return for which breaker type. But as I said this is just needed for unit tests, so I would rather provide a more generic implementation that allows this registration (for unit tests) and have `StaticCircuitBreakerService`(or `NoneCircuitBreakerService`) just use it internally.

bleskes replied:
> as I said this is just needed for unit tests

If people set the breaker type to none, this class is used and it's important it will return a NoopCircuitBreaker for each type.

PS. just double checking you know you can register an instance with Guice, not only classes.

danielmitterdorfer replied:
I found now an even simpler solution by using the real thing (i.e. `HierarchyCircuitBreakerService`) and set the breaker limit of the in-flights circuit breaker via settings. I'll revert the implementation of `StaticCircuitBreakerService` and also rename it again to `NoneCircuitBreakerService`.

> PS. just double checking you know you can register an instance with Guice, not only classes.

Yes, I am aware of that.

bleskes commented on core/src/main/java/org/elasticsearch/transport/local/LocalTransport.java:
```diff
@@ -288,6 +294,9 @@ private void handleRequest(StreamInput stream, long requestId, LocalTransport so
                 throw new ActionNotFoundTransportException("Action [" + action + "] not found");
             }
+            // reserve here in order be able to send a response in case we break the limit
+            inFlightRequestsBreaker.addEstimateBytesAndMaybeBreak(messageLengthBytes, "Node in-flight requests");
```
can the label be `<transport_request>` ? seems to be more consistent with other stuff.

danielmitterdorfer replied:
Sure. I adapt also the Netty layer and the HTTP layer accordingly.

bleskes commented on core/src/main/java/org/elasticsearch/transport/local/LocalTransport.java:
```diff
@@ -288,6 +294,9 @@ private void handleRequest(StreamInput stream, long requestId, LocalTransport so
             }
+            // reserve here in order be able to send a response in case we break the limit
+            inFlightRequestsBreaker.addEstimateBytesAndMaybeBreak(messageLengthBytes, "Node in-flight requests");
+            transportChannel.setReservedBytes(messageLengthBytes);
```
can we make this part of the constructor of the transportChannel? this also means we need to change the breaker earlier (which is fine because this is called under a try catch anyway.

danielmitterdorfer replied:
Well, in case the circuit breaker throws an exception in the constructor of the channel, we don't have any place where we can send an error response to the client (see the catch block of this method). If that is fine I can change that.

bleskes replied:
the whole handleRequest method is called under try catch

danielmitterdorfer replied:
As discussed, I'll reserve before creating the channel and have the enclosing try-catch handle `CircuitBreakingException`.

danielmitterdorfer replied:
I looked at this more closely. First of all it is possible to make both transport channel implementations immutable. However, in both transport cases the client does not get a response when the circuit breaker trips:

* LocalTransport: I think there is a problem in the catch block of `LocalTransport#messageReceived()`. If an exception is thrown, `transportServiceAdapter` does not have handler registered for the given request. The handler is just registered on the *sending* node but never on the *receiving* node (where the exception occurs). In the end, the exception is silently dropped and the client never gets the exception.
* NettyTransport: In this case, the exception ends up in `NettyTransport#exceptionCaught()` which is just closing the connection on exception and the client also never get hold of the exception.

So I propose that we implement handleRequest in both cases as follows (sketch):

```java
//imagine here local or netty transport
TransportChannel transportChannel = null; 
try {
  inFlightRequestsBreaker().addEstimateBytesAndMaybeBreak(messageLengthBytes, "<transport_request>");
  transportChannel = new TransportChannel(..., messageLengthBytes);
  // further request handling logic here
} catch (Throwable t) {
  // this can just happen if the circuit breaker tripped
  // create a transport channel which does not have to free any bytes
  if (transportChannel == null) {
  	transportChannel = new TransportChannel(..., 0);
  }
  transportChannel.sendResponse(e);
}
```

In that case we always send a response. Wdyt?

However, I think we have to check the case `LocalTransport#messageReceived()` (but I'd rather not do this in the scope of this ticket).

bleskes replied:
re LocalTransport: I think it's a bug over there. The error handling should be:

```
            if (sendRequestId != null) {
                TransportResponseHandler handler = sourceTransport.transportServiceAdapter.onResponseReceived(sendRequestId);
                if (handler != null) {
                    handleException(handler, new RemoteTransportException(nodeName(), localAddress, action, e));
                }
            } else {
                logger.warn("Failed to receive message for action [{}]", e, action);
            }
```

and it should also spawn the handleException to the workers of the source transport.


Re the netty implementation. I don't think a bug in our request handling code (i.e. a leaking exception) should lead disconnecting the node, which is sometimes good, sometimes bad. I'm good with your solution there.

bleskes commented on core/src/main/java/org/elasticsearch/transport/local/LocalTransportChannel.java:
```diff
@@ -98,17 +97,22 @@ public void sendResponse(Throwable error) throws IOException {
-        final byte[] data = stream.bytes().toBytes();
+    private void sendResponseData(byte[] data) {
+        sourceTransport.inFlightRequestsBreaker.addWithoutBreaking(-reservedBytes);
+        reservedBytes = 0;
```
can we make this send once? with same autoclosable logic suggested for the netty channel?

danielmitterdorfer replied:
Done.

bleskes commented on core/src/main/java/org/elasticsearch/transport/netty/NettyTransportChannel.java:
```diff
@@ -64,6 +65,10 @@ public NettyTransportChannel(NettyTransport transport, TransportServiceAdapter t
         this.profileName = profileName;
     }
 
+    public void setReservedBytes(long reservedBytes) {
```
can this be part of the constructor?

danielmitterdorfer replied:
Done.

bleskes commented on core/src/main/java/org/elasticsearch/transport/netty/NettyTransportChannel.java:
```diff
@@ -110,7 +116,9 @@ public void sendResponse(TransportResponse response, TransportResponseOptions op
         } finally {
-            if (!addedReleaseListener) {
+            transport.inFlightRequestsBreaker.addWithoutBreaking(-reservedBytes);
+            reservedBytes = 0;
```
can we use the same autocloseable suggestion?

danielmitterdorfer replied:
Done.

bleskes commented on core/src/test/java/org/elasticsearch/http/netty/NettyHttpChannelTests.java:
```diff
@@ -144,15 +147,18 @@ public void testThatAnyOriginWorks() {
         Settings settings = Settings.builder().build();
-        httpServerTransport = new NettyHttpServerTransport(settings, networkService, bigArrays, threadPool);
+        httpServerTransport = new NettyHttpServerTransport(settings, networkService, StaticCircuitBreakerService
+            .neverBreak(), bigArrays, threadPool);
```
please wrap on StaticCircuitBreakerService , so it will be easy to read

danielmitterdorfer replied:
Done.

bleskes commented on core/src/test/java/org/elasticsearch/http/netty/NettyHttpChannelTests.java:
```diff
@@ -144,15 +147,18 @@ public void testThatAnyOriginWorks() {
         NettyHttpRequest request = new NettyHttpRequest(httpRequest, writeCapturingChannel);
 
+        // reserve some arbitrary amount of bytes
+        int reservedBytes = 5;
```
do we want to randomize it to 0 every once in a while?

danielmitterdorfer replied:
Good idea. I've used `randomInt(5)` now.

bleskes commented on core/src/test/java/org/elasticsearch/http/netty/NettyHttpChannelTests.java:
```diff
@@ -169,16 +175,54 @@ public void testHeadersSet() {
         assertThat(response.headers().get(HttpHeaders.Names.CONTENT_TYPE), equalTo(resp.contentType()));
     }
 
+    public void testLimiterAbortsTooLargeRequests() {
```
I think the test needs to be renamed? also what are we testing here? it seems we call the breaking directly, so we're actually testing the breaker?

danielmitterdorfer replied:
You're right. After yesterday's changes this test does not make too much sense anymore. I'll remove it.

bleskes commented on core/src/test/java/org/elasticsearch/http/netty/NettyHttpChannelTests.java:
```diff
@@ -169,16 +175,54 @@ public void testHeadersSet() {
+        WriteCapturingChannel writeCapturingChannel = new WriteCapturingChannel();
+        NettyHttpRequest request = new NettyHttpRequest(httpRequest, writeCapturingChannel);
+
+        // reserve some arbitrary amount of bytes below the limit
```
it's not really arbitrary is it ? :)

danielmitterdorfer replied:
Now it is :) `final int reservedBytes = randomInt(limitBytes - 1);`

bleskes commented on core/src/test/java/org/elasticsearch/http/netty/NettyHttpChannelTests.java:
```diff
@@ -169,16 +175,54 @@ public void testHeadersSet() {
+        }
+    }
+
+    public void testReservationIsReturnedAfterResponding() {
```
new naming?

danielmitterdorfer replied:
changed to `testReservedBytesAreReturnedAfterResponding()`

bleskes commented on core/src/test/java/org/elasticsearch/http/netty/NettyHttpChannelTests.java:
```diff
@@ -414,5 +460,68 @@ public BytesReference content() {
+        private long count;
+
+        public static TestCircuitBreaker breakAlways() {
+            return withLimit(-1);
```
0 seems more intuitive to me (-1 means unset in many places)

danielmitterdorfer replied:
I've removed the method. It is not called anymore.

bleskes commented on core/src/test/java/org/elasticsearch/http/netty/NettyHttpChannelTests.java:
```diff
@@ -414,5 +460,68 @@ public BytesReference content() {
+
+    }
+
+    private static class TestCircuitBreaker implements CircuitBreaker {
```
I think we need to make all the methods synchronized here? we don't have any concurrency guarantees right now

danielmitterdorfer replied:
I don't think we need any concurrency guarantees here as the test is single-threaded.

bleskes replied:
OK for now, but I guaranteed someone will use this and be surprised. But sure, we can wait.

danielmitterdorfer replied:
As I now use `HierarchyCircuitBreakerService` in the test, `TestCircuitBreaker` is now also gone.

bleskes commented on core/src/test/java/org/elasticsearch/transport/netty/NettyTransportIT.java:
```diff
@@ -140,6 +142,8 @@ protected String handleRequest(Channel channel, StreamInput buffer, long request
                             if (reg == null) {
                                 throw new ActionNotFoundTransportException(action);
                             }
+                            transport.inFlightRequestsBreaker.addEstimateBytesAndMaybeBreak(messageLengthBytes, "Node in-flight requests");
```
if we push channel creation and circuit breaking to the parent, we can remove this here.

danielmitterdorfer replied:
This code went slightly out of date anyway over time and I tried to remove as much of it as possible.

bleskes commented on core/src/test/java/org/elasticsearch/transport/netty/NettyTransportRequestLimitIT.java:
```diff
@@ -0,0 +1,91 @@
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+
+@ClusterScope(scope = Scope.SUITE, minNumDataNodes = 1, maxNumDataNodes = 5, numClientNodes = 0)
+public class NettyTransportRequestLimitIT extends ESIntegTestCase {
```
why do we need a dedicated netty test? can we just test with random transports?

danielmitterdorfer replied:
I've moved this test to `CircuitBreakerServiceIT` but I am not sure whether we want it there. We have three possible client combinations that we could test: node client, transport client (local transport), transport client (netty transport). This test is irrelevant (and would also fail) with the node client so I had to prevent it (`@ClusterScope.transportClientRatio = 1`). 

If we had a possibility to randomize the transport (either local or netty) then it would make sense. The only possibilities I've found were explicit definition of a specific transport via system property or to prevent one of the two with `@ESIntegTestCase.SuppressLocalMode` or `@ESIntegTestCase.SuppressNetworkMode`. So I am not too sure whether it makes sense to move it to `CircuitBreakerServiceIT` in the end (or did I miss a possibility to randomize it except for adjusting settings manually?).

bleskes replied:
>  This test is irrelevant (and would also fail) with the node client so I had to prevent it (@ClusterScope.transportClientRatio = 1).

I think a better work around (and maybe even a better test) is to explicitly acquire a  client of one node (doesn't matter which type) and send the request to another node (for example, a node info request for a specific node, or a bulk for a single shard index, where you know where the shard is).


> If we had a possibility to randomize the transport (either local or netty) 

Our CI randomizes the transport implementation based on parameters (I think it was node.local=true/false , but it may have changed). Using it means you don't have to do all the hary stuff like configuring ports and such.

bleskes commented:
Left a bunch of comments, but I think this looks great. Maybe I missed it , but do we have a test for the limit enforcing on the http layer?

danielmitterdorfer commented:
I just had unit tests on HTTP layer. I've also added an integration test. I've also adjusted the behavior on HTTP level. Instead of closing the connection silently, we return the associated status of `CircuitBreakingException` (which is now 500).

danielmitterdorfer commented:
@bleskes I add another round of commits which address all of your outstanding comments. Would be great if you could have another look.

bleskes commented on core/src/main/java/org/elasticsearch/transport/local/LocalTransportChannel.java:
```diff
@@ -102,16 +101,28 @@ public void sendResponse(Throwable error) throws IOException {
-                targetTransport.messageReceived(data, action, sourceTransport, version, null);
-            }
-        });
+        if (closed.get()) {
```
I think we can just call `close();` here and free the bytes before we call the execute on the targetTransport

danielmitterdorfer replied:
Adjusted accordingly in `LocalTransportChannel` and `NettyTransportChannel`.

bleskes commented on core/src/main/java/org/elasticsearch/transport/netty/NettyTransportChannel.java:
```diff
@@ -86,6 +85,7 @@ public void sendResponse(TransportResponse response) throws IOException {
 
     @Override
     public void sendResponse(TransportResponse response, TransportResponseOptions options) throws IOException {
+        checkChannelOpen();
```
same comment - just call close here...

bleskes commented on core/src/main/java/org/elasticsearch/transport/netty/NettyTransportChannel.java:
```diff
@@ -116,16 +116,17 @@ public void sendResponse(TransportResponse response, TransportResponseOptions op
 
     @Override
     public void sendResponse(Throwable error) throws IOException {
+        checkChannelOpen();
```
same about call close

bleskes commented on core/src/test/java/org/elasticsearch/indices/memory/breaker/CircuitBreakerServiceIT.java:
```diff
@@ -337,4 +351,69 @@ public void testCustomCircuitBreakerRegistration() throws Exception {
+            .put(HierarchyCircuitBreakerService.IN_FLIGHT_REQUESTS_CIRCUIT_BREAKER_LIMIT_SETTING.getKey(), inFlightRequestsLimit)
+            .build();
+
+        assertAcked(client().admin().cluster().prepareUpdateSettings().setTransientSettings(limitSettings));
```
do we want to do this just before indexing? to make sure we don't block anything else?

danielmitterdorfer replied:
I missed that one. It's now moved right before we start indexing.

bleskes commented on core/src/main/java/org/elasticsearch/common/breaker/CircuitBreakingException.java:
```diff
@@ -67,6 +68,11 @@ public long getByteLimit() {
 
     @Override
+    public RestStatus status() {
+        return RestStatus.INTERNAL_SERVER_ERROR;
```
do we need this? it's like the base class (as we never have a cause)

danielmitterdorfer replied:
I think this is justified as we want to enforce 500 no matter what. Otherwise we can run into the fragile base class problem. 

If we really want to remove the override, we have at least `NettyHttpRequestSizeLimitIT` as a safety guard which will break if the status code changes.

bleskes replied:
To me I think the rule should be: we override this method if you have a specific rest http status code that's associated with the exception. I this case we just use a generic code that's not really indicative of the error. However, looking at alternative, I suggest we use 503 Service Unavailable

danielmitterdorfer replied:
I am also fine with this status code. I've changed it accordingly.

bleskes commented on core/src/main/java/org/elasticsearch/http/HttpServer.java:
```diff
@@ -98,7 +100,12 @@ public void dispatchRequest(HttpRequest request, HttpChannel channel, ThreadCont
+        try {
+            transport.addBytes(request);
+            restController.dispatchRequest(request, new ResourceHandlingHttpChannel(channel, transport), threadContext);
+        } catch (CircuitBreakingException ex) {
```
we can make this catch throwable and remove the catch from https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/rest/RestController.java#L175

danielmitterdorfer replied:
This is very subtle, but I think we cannot do this. `HttpServer` will send the error response on the raw Netty channel wherea `RestController` will send the error response on `ResourceHandlingHttpChannel`. In the former case no bytes will be freed (as no bytes have been reserved) and in the latter case we ensure that all reserved bytes are freed again. 

But I can still move all exception handling logic into `HttpServer`. I just have to keep track via which channel I send an error response. Something along these lines should do it:

```java
HttpChannel responseChannel = channel;
try {
  transport.addBytes(request);
  responseChannel = new ResourceHandlingHttpChannel(channel, transport);
  restController.dispatchRequest(request, responseChannel, threadContext);
} catch (Throwable t) {
  restController.sendErrorResponse(request, responseChannel, t);
}
```

bleskes replied:
++ the the plan to move it here.

danielmitterdorfer replied:
I've moved it now.

bleskes commented on core/src/main/java/org/elasticsearch/http/HttpServer.java:
```diff
@@ -42,25 +43,18 @@
         this.transport = transport;
         this.restController = restController;
         this.nodeService = nodeService;
-        nodeService.setHttpServer(this);
```
why this change?

danielmitterdorfer replied:
As I created `HttpServerTests` I noticed that `HttpServer` has a mutual dependency on `NodeService` which is a very heavyweight dependency and made it hard to test `HttpServer`. I did not want to get rid of this dependency within the scope of this PR but still reduce the coupling a bit. That's why I removed this line and let Guice inject the HttpServer by adding `setHttpServer` (similarly to the already existing method `setScriptService`). After this PR is done I want to look more closely whether I can get rid of this hard mutual dependency alltogether. And finally: Wiring code like this line belongs to infrastructure code and clutters domain classes.

bleskes replied:
I see and I understand the reasoning. However, Guice circular dependency resolution is very tricky and if Guice can't resolve things correctly you get the most horrific, cryptic, MB long errors. This is why we are slowing removing Guice usage. I suspect there was a good reason as to why this is done this way and I rather not change this as part of this PR. Can we do that as part of the follow up too?

danielmitterdorfer replied:
I'll undo the change. As a consequence, the setup code in `HttpServerTest` will get more complex.

bleskes commented on core/src/main/java/org/elasticsearch/http/HttpServer.java:
```diff
@@ -71,15 +65,23 @@ protected void doStart() {
         if (logger.isInfoEnabled()) {
             logger.info("{}", transport.boundAddress());
         }
+        registerHttpAddress();
```
why is this needed? calling a protected method from the constructor isn't considered a good practice?

danielmitterdorfer replied:
I agree with your conclusion but not with your premise. This is a lifecycle method and not a constructor.

bleskes replied:
argh. Missed the grey method start line in the github diff.

bleskes commented on core/src/main/java/org/elasticsearch/http/HttpServer.java:
```diff
@@ -117,4 +124,24 @@ void handleFavicon(HttpRequest request, HttpChannel channel) {
+        private final HttpServerTransport transport;
+
+        public ResourceHandlingHttpChannel(HttpChannel delegate, HttpServerTransport transport) {
+            super(delegate.request(), delegate.detailedErrorsEnabled());
```
this is not good...  you rely on the base class implementation (that only these two parameters for it's state). You have to proxy all requests to the delegate

danielmitterdorfer replied:
You're right! When we think about API evolution I think it is dangerous that `ResourceHandlingHttpChannel` extends from a class (`HttpChannel`) instead of implementing an interface. I think introducing an interface for `RestChannel` (or turning `RestChannel` itself into an interface) is a good longer term solution as we need to enforce that we delegate all methods and when the interface changes we get compile errors. If we stick with `RestChannel` as a base class we can easily add a new method to the base class and miss a delegation. Wdyt?

bleskes replied:
I agree this is tricky. I would suggest making the RestChannel an interface it self and move the logic to the HttpChannel, which can be an abstract class. If you agree, can we make this little change in a different PR, get that in and merge it into this branch? This thing is getting bigger so any change we can do in isolation, we should..

danielmitterdorfer replied:
I'll do this but due to package dependencies I'd create the base class in the `rest` package. We sometimes override `RestChannel` in tests and I think it makes sense to keep this. However, I will get rid of `HttpChannel` and related overhead classes then.

bleskes replied:
++

bleskes commented on core/src/main/java/org/elasticsearch/http/HttpServer.java:
```diff
@@ -117,4 +124,24 @@ void handleFavicon(HttpRequest request, HttpChannel channel) {
+
+        @Override
+        public void sendResponse(RestResponse response) {
+            try {
```
can we have the same close patteren?

danielmitterdorfer replied:
Adjusted.

bleskes commented on core/src/main/java/org/elasticsearch/http/HttpServerTransport.java:
```diff
@@ -34,4 +36,8 @@
 
     void httpServerAdapter(HttpServerAdapter httpServerAdapter);
+
+    void addBytes(RestRequest request) throws CircuitBreakingException;
```
can we call this addRequestBytes/freeRequestBytes and add java docs?

danielmitterdorfer replied:
Done.

bleskes commented on core/src/main/java/org/elasticsearch/http/netty/NettyHttpChannel.java:
```diff
@@ -85,6 +88,9 @@ public BytesStreamOutput newBytesOutput() {
 
     @Override
     public void sendResponse(RestResponse response) {
+        if (closed.get()) {
```
we don't really need this here any more.. (it needs to be higher up)

danielmitterdorfer replied:
Ah, that's a leftover. I remove it as the logic is now contained in `ResourceHandlingHttpChannel`.

bleskes commented on core/src/main/java/org/elasticsearch/http/netty/NettyHttpServerTransport.java:
```diff
@@ -274,7 +269,7 @@ public NettyHttpServerTransport(Settings settings, NetworkService networkService
         }
         this.maxContentLength = maxContentLength;
-        this.inFlightRequestsBreaker = circuitBreakerService.getBreaker(CircuitBreaker.IN_FLIGHT_REQUESTS);
+        this.circuitBreakerService = circuitBreakerService;
```
why not move all of this to the HttpServer? then we don't need to change the interface and keep things local..

danielmitterdorfer replied:
Done.

bleskes commented on core/src/test/java/org/elasticsearch/http/HttpServerTests.java:
```diff
@@ -0,0 +1,254 @@
+        assertEquals(0, httpServerTransport.reservedBytes());
+    }
+
+    public void testDispatchRequestAddsAndFreesBytesOnlyOnceOnError() {
```
you need to generate two errors for this to test what you want :)

danielmitterdorfer replied:
Actually I wanted to test that in case an error is thrown in a `RestHandler` (see the registered handlers in `#setup()`), we still release bytes only once when sending an error response via the `RestController`. So I think this is ok.

bleskes replied:
OK. I though you want to check duplicate exceptions (like an exception when handling an exception) which is a good idea. I thought you you mean to do this because the test has OnlyOnce in the name as a opposed to `testDispatchRequestAddsAndFreesBytesOnError`

danielmitterdorfer replied:
I've corrected the test name and also added another one that does what you thought originally. :)

bleskes commented on core/src/main/java/org/elasticsearch/transport/netty/MessageChannelHandler.java:
```diff
@@ -285,6 +261,11 @@ protected String handleRequest(Channel channel, StreamInput buffer, long request
     }
 
+    protected void validateRequest(Marker marker, StreamInput buffer, long requestId, String action) throws IOException {
+        marker.validateRequest(buffer, requestId, action);
```
I meant take the marker logic and put it in two methods :) this way it is a pretty useless method...

danielmitterdorfer replied:
By itself, the method is useless, right. But previously `NettyTransportIT`had a custom `MessageChannelHandler` implementation which had to reimplement `#handleRequest()` almost completely just to inject custom errors in the test (and obviously the implementation already deviated from production code). By introducting this method, it is possible to override request validation much more fine-grained and still inject custom error detection in test code at the right place.

bleskes replied:
OK. Fair enough. Can we add a comment?

danielmitterdorfer replied:
Sure. I'll add one.

bleskes commented on core/src/main/java/org/elasticsearch/transport/netty/NettyTransport.java:
```diff
@@ -308,6 +312,10 @@ ThreadPool threadPool() {
         return threadPool;
     }
 
+    CircuitBreaker inFlightRequestsBreaker() {
```
+1 on the getting method, but why not resolve the breaker at the constructor?

danielmitterdorfer replied:
The circuit breaker limit can be changed dynamically and the circuit breaker service creates a new instance when the setting changes. That's why we always have to get a fresh breaker. Otherwise it would not reflect the changed settings.

bleskes replied:
good catch. Can you add a comment?

danielmitterdorfer replied:
Done.

bleskes commented:
@danielmitterdorfer look at the new implementation. Feels better. I added some cleanup comments...

danielmitterdorfer commented:
@bleskes Thanks again. I have aligned the close logic and moved HTTP request limiting to HttpServer. I answered all other comments but did not change the code yet.

bleskes commented on core/src/main/java/org/elasticsearch/http/HttpServer.java:
```diff
@@ -100,11 +99,14 @@ public void dispatchRequest(HttpRequest request, HttpChannel channel, ThreadCont
+            // iff we could reserve bytes for the request we need to send the response also over this channel
+            responseChannel = new ResourceHandlingHttpChannel(channel, circuitBreakerService);
+            restController.dispatchRequest(request, responseChannel, threadContext);
+        } catch (Throwable t) {
```
can we clean up the other try catch? it's not needed now (as we catch things here too)

danielmitterdorfer replied:
I've already removed the try-catch block from `RestController#executeHandler()`. Do you refer to another try-catch?

bleskes replied:
I meant that one indeed. GitHub is playing tricks - it didn't show me that change in the same commit (but now it does). Sorry for the noise.

danielmitterdorfer commented:
@bleskes All your review comments should be addressed with the last round of commits now.

bleskes commented:
Code LGTM! Thx @danielmitterdorfer . We still have the open question of what should be a good default value for this breaker. I feel we have two options:

1) Set it to some percentage of total node memory - the complicated thing here is that different nodes have different requirements. For example, coordinating only nodes can have a lot of inflight requests (this is their job) but data nodes may have different requirements. 
2) Don't set a max and allow the parent circuit breaker to trigger where the sum of request + whatever the other breakers (field data etc.) need reaches above an accepted limit. 

@dakrone @clintongormley do you have any opinions here?

danielmitterdorfer commented:
@bleskes Thanks for your review. I have now introduced constants for all Netty headers and added Javadocs and additional helper methods in `Marker`. It should now be quite obvious what is read w.r.t. to headers. I felt that it may also make sense to add additional methods in `NettyHeaders` that help with low-level header handling (like reading the marker bytes) but I did not do this in this PR. However, this might be worth doing in another small PR (later).

bleskes commented on core/src/main/java/org/elasticsearch/transport/netty/SizeHeaderFrameDecoder.java:
```diff
@@ -41,7 +41,8 @@
     @Override
     protected Object decode(ChannelHandlerContext ctx, Channel channel, ChannelBuffer buffer) throws Exception {
-        if (buffer.readableBytes() < 6) {
+        int sizeHeaderLength = NettyHeader.MARKER_BYTES_SIZE + NettyHeader.MESSAGE_LENGTH_SIZE;
```
make this final?

danielmitterdorfer replied:
Done.

bleskes commented:
sounds good @danielmitterdorfer

dakrone commented:
> We still have the open question of what should be a good default value for this breaker. I feel we have two options:
> 1) Set it to some percentage of total node memory - the complicated thing here is that different nodes have different requirements. For example, coordinating only nodes can have a lot of inflight requests (this is their job) but data nodes may have different requirements.
> 2) Don't set a max and allow the parent circuit breaker to trigger where the sum of request + whatever the other breakers (field data etc.) need reaches above an accepted limit.

> @dakrone @clintongormley do you have any opinions here?

Hmm... I think setting it to 200mb sounds reasonable to me? These are nice dynamic settings, so it's easy to change at runtime and I'd rather err too low than too high.

dakrone commented:
@bleskes brought up that this is all requests, so 200mb is probably too small (and I agree), since the parent breaker limit is 70%, how about setting this to 60%?

danielmitterdorfer commented:
@dakrone: The suggestion with 60% is fine for me. If I hear no objections I change it accordingly and update the docs. /cc:@bleskes

danielmitterdorfer commented:
We have decided to not limit the respective circuit breaker but rely on the parent circuit breaker (which is currently set to 70%). I'll update code and docs accordingly and merge the PR afterwards.

jasontedor commented on core/src/main/java/org/elasticsearch/transport/netty/NettyTransportChannel.java:
```diff
@@ -134,6 +141,14 @@ public void sendResponse(Throwable error) throws IOException {
         transportServiceAdapter.onResponseSent(requestId, action, error);
     }
 
+    private void close() {
```
@danielmitterdorfer Why is this method called `close`? It might make more sense if it were in the `finally` blocks but as-is I'm confused by it. Help?

danielmitterdorfer replied:
In my original implementation it was called indeed in the `finally` block, but Boaz and I had a lengthy discussion about this one. It makes sense to free the reserved bytes at the beginning of the method as we're done with request processing at this point in time. Freeing at the beginning guards against any problems that might occur further down the line (admittedly this could also be solved in a finally block).

The name `close()` just implies that you cannot reuse the channel after this method has been called and the precondition check guards against this.

jasontedor replied:
I think it should be in the finally block, regardless of the name. We asynchronously write, the bytes are still held until after the write. :smile:

If you've already had that discussion and concluded that at the beginning of the method is correct (and unswayed by my thoughts here), can we at least change the name? The semantics are just counterintuitive and a name change would at least improve that. Sorry, I do not have any good suggestions. :frown:

bleskes replied:
> We asynchronously write, the bytes are still held until after the write. 

I don't follow this one? this is about freeing the bytes of the *request* not the bytes of the response? it's all just a heuristic anyway.

I'm good with doing it in a finally clause, though I think close() indicates well that the channel is closed and shouldn't be used again. If we agree on that I think that closing at the beginning and throwing an early exception if already closed is what we want?

jasontedor replied:
> I don't follow this one? this is about freeing the bytes of the request not the bytes of the response?

Yes, I'm sorry, I confused myself and brought something irrelevant into this discussion.

> I'm good with doing it in a finally clause, though I think close() indicates well that the channel is closed and shouldn't be used again.

That's why I don't like it, we are "closing" and then still proceeding to perform operations on the channel. I don't like the semantics. Can I suggest `release` or `unreserve`?

bleskes replied:
well, we don't really do ops on the `NettyTransportChannel` but rather it's underlying fields  which is similar in my mind to any clean up on close code. But I'm with you if you want to change this - I personally find release/unreserve even more confusing :)

danielmitterdorfer replied:
>  we are "closing" and then still proceeding to perform operations on the channel.

That's a pure implementation detail. The caller cannot observe this. It just observes that after the response is sent once, it cannot perform any "write" operation (sending a response) again.

> Can I suggest release or unreserve?

From the perspective of the channel it is still "closed", but if we really have to do this, then I'd more lean towards `release`.

jasontedor replied:
> That's a pure implementation detail.

That's exactly what we are discussing here though? The semantics of *this* implementation.

> From the perspective of the channel it is still "closed", but if we really have to do this, then I'd more lean towards release.

This was never an issue though? It was added as part of this PR to release the tracked bytes (also added in this PR).

danielmitterdorfer replied:
I still think that `close()` is fine as name but I am ok if you want to change it.

jasontedor replied:
Thanks @danielmitterdorfer, I pushed 86f1bedaab683d77b86daa92d7ddde5aee4cedd7.
